File: iOS/Sources/Data/Repositories/InMemoryAuthRepository.swift
import Foundation

actor InMemoryAuthRepository: AuthRepositoryProtocol {
    private var users: [String: (User, String)] = [:] // email -> (user, passwordHash)
    private var storedToken: String?
    private var currentUser: User?
    private var nextId: Int64 = 1

    func register(email: String, password: String, name: String) async throws -> AuthResponse {
        if users[email] != nil {
            throw DomainError.emailAlreadyExists
        }

        let user = User(
            id: nextId,
            email: email,
            name: name,
            createdAt: Date(),
            updatedAt: Date()
        )
        nextId += 1

        users[email] = (user, String(password.hashValue))

        let token = "mock_token_\(user.id)_\(Int(Date().timeIntervalSince1970))"
        storedToken = token
        currentUser = user

        return AuthResponse(token: token, user: user)
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        guard let (user, storedPasswordHash) = users[email] else {
            throw DomainError.invalidCredentials
        }

        if storedPasswordHash != String(password.hashValue) {
            throw DomainError.invalidCredentials
        }

        let token = "mock_token_\(user.id)_\(Int(Date().timeIntervalSince1970))"
        storedToken = token
        currentUser = user

        return AuthResponse(token: token, user: user)
    }

    func getStoredToken() async -> String? {
        return storedToken
    }

    func saveToken(_ token: String) async {
        storedToken = token
    }

    func clearToken() async {
        storedToken = nil
    }

    func getCurrentUser() async -> User? {
        return currentUser
    }

    func saveCurrentUser(_ user: User) async {
        currentUser = user
    }

    func clearCurrentUser() async {
        currentUser = nil
    }
}

File: android/app/src/main/java/com/budgetmanager/data/repository/InMemoryAuthRepository.kt
package com.budgetmanager.data.repository

import com.budgetmanager.domain.entity.AuthResponse
import com.budgetmanager.domain.entity.User
import com.budgetmanager.domain.repository.AuthRepository
import java.time.LocalDateTime

class InMemoryAuthRepository : AuthRepository {
    private val users = mutableMapOf<String, Pair<User, String>>() // email -> (user, passwordHash)
    private var storedToken: String? = null
    private var currentUser: User? = null
    private var nextId: Long = 1

    override suspend fun register(email: String, password: String, name: String): Result<AuthResponse> {
        if (users.containsKey(email)) {
            return Result.failure(Exception("Email already exists"))
        }

        val user = User(
            id = nextId++,
            email = email,
            name = name,
            createdAt = LocalDateTime.now(),
            updatedAt = LocalDateTime.now()
        )
        users[email] = Pair(user, password.hashCode().toString())

        val token = "mock_token_${user.id}_${System.currentTimeMillis()}"
        storedToken = token
        currentUser = user

        return Result.success(AuthResponse(token = token, user = user))
    }

    override suspend fun login(email: String, password: String): Result<AuthResponse> {
        val userPair = users[email]
            ?: return Result.failure(Exception("Invalid email or password"))

        val (user, storedPasswordHash) = userPair
        if (storedPasswordHash != password.hashCode().toString()) {
            return Result.failure(Exception("Invalid email or password"))
        }

        val token = "mock_token_${user.id}_${System.currentTimeMillis()}"
        storedToken = token
        currentUser = user

        return Result.success(AuthResponse(token = token, user = user))
    }

    override suspend fun getStoredToken(): String? = storedToken

    override suspend fun saveToken(token: String) {
        storedToken = token
    }

    override suspend fun clearToken() {
        storedToken = null
    }

    override suspend fun getCurrentUser(): User? = currentUser

    override suspend fun saveCurrentUser(user: User) {
        currentUser = user
    }

    override suspend fun clearCurrentUser() {
        currentUser = null
    }
}
